cmake/bintools/arcmwdt/target.cmake

```diff
@@ -0,0 +1,371 @@
+#      RESULT_BYPROD_LIST <List of command output byproducts>
+#
+#      STRIP_ALL          <When present, remove relocation and symbol info>
+#      STRIP_DEBUG        <When present, remove debugging symbols and sections>
+#
+#      TARGET_INPUT       <Input file format type>
+#      TARGET_OUTPUT      <Output file format type>
+#
+#      GAP_FILL           <Value used for gap fill, empty or not set, no gap fill>
+#      SREC_LEN           <For srec format only, max length of the records>
+#
+#      SECTION_ONLY       <One or more section names to be included>
+#      SECTION_REMOVE     <One or more section names to be excluded>
+#      SECTION_RENAME     <One or more section names to be renamed 'from=to'>
+#
+#      FILE_INPUT         <The input file>
+#      FILE_OUTPUT        <The output file>
+#    )
+function(bintools_objcopy)
+  cmake_parse_arguments(
+    # Prefix of output variables
+    BINTOOLS_OBJCOPY
+    # List of argument names without values, hence boolean
+    "STRIP_ALL;STRIP_DEBUG"
+    # List of argument names with one value
+    "RESULT_CMD_LIST;RESULT_BYPROD_LIST;TARGET_INPUT;TARGET_OUTPUT;GAP_FILL;SREC_LEN;FILE_INPUT;FILE_OUTPUT"
+    # List of argument names with multible values
+    "SECTION_ONLY;SECTION_RENAME;SECTION_REMOVE"
+    # Parser input
+    ${ARGN}
+  )
+
+  # Verify arguments
+  if(NOT DEFINED BINTOOLS_OBJCOPY_RESULT_CMD_LIST OR NOT DEFINED ${BINTOOLS_OBJCOPY_RESULT_CMD_LIST})
+    message(FATAL_ERROR "RESULT_CMD_LIST is required.")
+  elseif(NOT DEFINED BINTOOLS_OBJCOPY_FILE_INPUT OR NOT DEFINED BINTOOLS_OBJCOPY_FILE_OUTPUT)
+    message(FATAL_ERROR "Both FILE_INPUT and FILE_OUTPUT is required.")
+  endif()
+
+  # Handle stripping
+  if (DEFINED BINTOOLS_OBJCOPY_STRIP OR DEFINED BINTOOLS_OBJCOPY_STRIP_ALL)
+    if(${BINTOOLS_OBJCOPY_STRIP_ALL})
+      set(obj_copy_strip "-qs")
+    elseif(${BINTOOLS_OBJCOPY_STRIP_DEBUG})
+      set(obj_copy_strip "-ql")
+    endif()
+    set(obj_copy_cmd
+        COMMAND ${CMAKE_STRIP} ${obj_copy_strip}
+        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
+  endif()
+
+  # no support of --srec-len in mwdt
+
+  # Handle Input and Output target types
+  if(DEFINED BINTOOLS_OBJCOPY_TARGET_OUTPUT)
+    set(obj_copy_target_output "")
+    set(obj_copy_gap_fill "")
+    if(DEFINED BINTOOLS_OBJCOPY_GAP_FILL)
+    set(obj_copy_gap_fill "-f;${BINTOOLS_OBJCOPY_GAP_FILL}")
+    endif()
+      # only mwdt's elf2hex supports gap fill
+    if(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "srec")
+      set(obj_copy_target_output "-m")
+    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "ihex")
+      set(obj_copy_target_output "-I")
+    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "binary")
+      set(obj_copy_target_output "-B")
+    endif()
+     set(obj_copy_cmd
+        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
+        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
```

Comment:
This will not work as expected.
When this macro is being called from here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L1134-L1142
then it is called with `STRIP_ALL` and `TARGET_OUTPUT ihex`, which will result in first, line 132-134:
```
   set(obj_copy_cmd
       COMMAND ${CMAKE_STRIP} ${obj_copy_strip}
        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
```
which will take the `FILE_INPUT` and create `FILE_OUTPUT`.
But then at line 154-156, this happens:
```
     set(obj_copy_cmd
        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
```
which basically means that the original `FILE_INPUT` is used again and then overwriting `FILE_OUTPUT` which effectively will result in the `STRIP_ALL` to have no effect.

Reply:
Doesn't elf2hex strip out any debug info anyway, leaving only the binary code and no debug sections?

Reply:
Good catch! So that basically happens because in existing case with GNU toolchain we do both manipulations in 1 go here https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-mwdt/cmake/bintools/gnu/target_objcopy.cmake#L111:
```cmake
  # Construct the command
  set(obj_copy_cmd
    # Base command
    COMMAND ${CMAKE_OBJCOPY} ${obj_copy_strip} ${obj_copy_gap_fill} ${obj_copy_srec_len}
    # Input and Output target types
    ${obj_copy_target_input} ${obj_copy_target_output}
    # Sections
    ${obj_copy_sections_only} ${obj_copy_sections_rename} ${obj_copy_sections_remove}
    # Input and output files
    ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT}
  )
```
while in case of MW tools we don't have a direct substitution of really powerful `objcopy` and have a multi-stage pipeline of binary processing, thus separate calls to `stripac`, `elf2hex` etc.

Moreover [though I'm far from being even an experienced user of CMake] I'm not sure if all those manipulations are being done since we do `set(obj_copy_cmd ...` multiple times... I'd think that way we **re-define** the same command again and again. Correct me if I'm wrong though.
```
  # Handle Input and Output target types
  if(DEFINED BINTOOLS_OBJCOPY_TARGET_OUTPUT)
    set(obj_copy_target_output "")
    set(obj_copy_gap_fill "")
    if(DEFINED BINTOOLS_OBJCOPY_GAP_FILL)
    set(obj_copy_gap_fill "-f;${BINTOOLS_OBJCOPY_GAP_FILL}")
    endif()
      # only mwdt's elf2hex supports gap fill
    if(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "srec")
      set(obj_copy_target_output "-m")
    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "ihex")
      set(obj_copy_target_output "-I")
    elseif(${BINTOOLS_OBJCOPY_TARGET_OUTPUT} STREQUAL "binary")
      set(obj_copy_target_output "-B")
    endif()
     set(obj_copy_cmd                                           <-- *** HERE ***
        COMMAND ${CMAKE_ELF2HEX} ${obj_copy_gap_fill} ${obj_copy_target_output}
        -o ${BINTOOLS_OBJCOPY_FILE_OUTPUT} ${BINTOOLS_OBJCOPY_FILE_INPUT})
  endif()

  # Handle sections, if any
  # 1. Section only selection(s)
  set(obj_copy_sections_only "")
  if(DEFINED BINTOOLS_OBJCOPY_SECTION_ONLY)
    foreach(section_only ${BINTOOLS_OBJCOPY_SECTION_ONLY})
      list(APPEND obj_copy_sections_only "-sn;${section_only}")
    endforeach()
    set(obj_copy_cmd                                           <-- *** HERE ***
        COMMAND ${CMAKE_ELF2BIN} -q ${obj_copy_sections_only}
        ${BINTOOLS_OBJCOPY_FILE_INPUT} ${BINTOOLS_OBJCOPY_FILE_OUTPUT})
  endif()
```

Still given we don't currently have any ARC boards supported by upstream Zephyr which are programmed with binaries (we load Elf contents with either OpenOCD or MDB via JTAG) this doesn't affect `sanitychecker` results and given current race to the feature freeze I guess we may address it via formal bug-fix during stabilization period. Agree?

Reply:
> (we load Elf contents with either OpenOCD or MDB via JTAG)

agreed, then we can take it in, as-is.

> set(obj_copy_cmd ...

right, the last `set` wins.
(it should have been `list(APPEND ...)` in order to work.)